etp-core/etp-backend/src/main/clj/solita/etp/service/perusparannuspassi.clj

```diff
@@ -173,39 +222,46 @@
         ;; find-perusparannuspassi already only looks for owned ppps.
         (assert-update-requirements! tx whoami current-ppp ppp)
 
-        (db/with-db-exception-translation jdbc/update! tx :perusparannuspassi
-                                          (-> ppp
-                                              (dissoc :energiatodistus-id :vaiheet)
-                                              ppp->db-row)
-                                          ["id = ?" id]
-                                          db/default-opts)
-
-        (doseq [vaihe (:vaiheet ppp)]
-          (db/with-db-exception-translation jdbc/update! tx :perusparannuspassi-vaihe
-                                            (-> vaihe
-                                                (dissoc :vaihe-nro :perusparannuspassi-id)
-                                                without-toimenpide-ehdotukset
-                                                ppp-vaihe->db-row)
-                                            ["perusparannuspassi_id = ? and vaihe_nro = ?" id (:vaihe-nro vaihe)]
+        (let [ppp-db-row (-> ppp
+                             (dissoc :energiatodistus-id :vaiheet)
+                             ppp->db-row)
+              ;; Validate PPP main row numeric values (errors throw, warnings collected)
+              warnings (validate-ppp-db-row! tx ppp-db-row 2026)]
+
+          ;; Update the main PPP row
+          (db/with-db-exception-translation jdbc/update! tx :perusparannuspassi
+                                            ppp-db-row
+                                            ["id = ?" id]
                                             db/default-opts)
-          (jdbc/delete! tx :perusparannuspassi_vaihe_toimenpide_ehdotus
-                        ["perusparannuspassi_id = ? and vaihe_nro = ?" id (:vaihe-nro vaihe)]
-                        db/default-opts)
-          (doseq [[ordinal toimenpide-ehdotus-id]
-                  (map vector (range) (->> vaihe :toimenpiteet :toimenpide-ehdotukset (map :id)))]
-            (db/with-db-exception-translation
-              jdbc/insert! tx :perusparannuspassi_vaihe_toimenpide_ehdotus
-              {:perusparannuspassi_id id
-               :vaihe_nro             (:vaihe-nro vaihe)
-               :toimenpide_ehdotus_id toimenpide-ehdotus-id
-               :ordinal               ordinal}
-              db/default-opts)))
-        (doseq [vaihe-nro (clojure.set/difference #{1 2 3 4} (->> ppp :vaiheet (map :vaihe-nro)))]
-          (jdbc/update! tx :perusparannuspassi-vaihe
-                        {:valid false}
-                        ["perusparannuspassi_id = ? and vaihe_nro = ?" id vaihe-nro]
-                        db/default-opts))
-        nil)
+
+          ;; Update the vaihe rows
```

Comment:
Vaiheen validointi vielä aikalailla puuttuu, mutta ei myöskään ole oikein mielekkäitä raja-arvoja